Isolated result spaces broken when not extending

### System Info
- Operating System: Ubuntu, OS X
- Python Version: `python --version`
- Version of catkin_tools: 0.4.3
- ROS Distro: Indigo
### Build / Run Issue

The following sequence results in an error:

```
$ rosinstall_generator --rosdistro indigo roscpp --deps --tar > roscpp.rosinstall
$ rosinstall src roscpp.rosinstall -j8
$ catkin config --isolate-install --isolate-devel --install  # also fails with --no-install
$ catkin build  # also fails with -p1
[ snip ]
Starting  >>> roscpp_serialization
___________________________________________________________________________________
Errors     << roscpp_serialization:cmake 
CMake Error at /Users/mikepurvis/isolate_install_experiments/devel/catkin/share/catkin/cmake/catkinConfig.cmake:83 (find_package):
  Could not find a package configuration file provided by "roscpp_traits"
  with any of the following names:

    roscpp_traitsConfig.cmake
    roscpp_traits-config.cmake

  Add the installation prefix of "roscpp_traits" to CMAKE_PREFIX_PATH or set
  "roscpp_traits_DIR" to a directory containing one of the above files.  If
  "roscpp_traits" provides a separate development package or SDK, be sure it
  has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:3 (find_package)
```

The following variants in the same workspace run to completion with a successful result:

```
src/catkin/bin/catkin_make_isolated
src/catkin/bin/catkin_make_isolated --install
src/catkin/bin/catkin_make_isolated --install --merge
some/other/workspace/env.sh catkin build
```

Not terribly clear what is going on here, but it seems that somehow the environment in the isolated build case isn't quite getting populated correctly.

--- a/catkin_tools/jobs/catkin.py
+++ b/catkin_tools/jobs/catkin.py
@@ -380,7 +380,7 @@ def create_catkin_build_job(context, package, package_path, dependencies, force_
     stages.append(FunctionStage(
         'loadenv',
         loadenv,
-        locked_resource='installspace',
+        locked_resource=None if context.isolate_install else 'installspace',
         job_env=job_env,
         package=package,
         context=context
@@ -504,7 +504,7 @@ def create_catkin_build_job(context, package, package_path, dependencies, force_
             [MAKE_EXEC, 'install'],
             cwd=build_space,
             logger_factory=CMakeMakeIOBufferProtocol.factory,
-            locked_resource='installspace'
+            locked_resource=None if context.isolate_install else 'installspace'
         ))
 
     return Job(

--- a/catkin_tools/jobs/utils.py
+++ b/catkin_tools/jobs/utils.py
@@ -57,6 +57,11 @@ def loadenv(logger, event_queue, job_env, package, context):
     if context.destdir is not None:
         job_env['_CATKIN_SETUP_DIR'] = context.package_dest_path(package)
 
+    updated_env = {}
+
+    # Get the envvars associated with each workspace. Order is unimportant here since
+    # there will either be only one env loader path (linked/merged result space) or
+    # there will be many, but dependencies don't have a fixed order.
     for env_loader_path in env_loader_paths:
         if logger:
             logger.out('Loading environment from: {}'.format(env_loader_path))
@@ -66,8 +71,20 @@ def loadenv(logger, event_queue, job_env, package, context):
             quiet=True,
             cached=context.use_env_cache,
             strict=False)
-        job_env.update(resultspace_env)
+        for key, values_str in resultspace_env.items():
+            if key in updated_env:
+                updated_env[key] |= set(values_str.split(':'))
+            else:
+                updated_env[key] = set(values_str.split(':'))
 
+    # Now merge the updated environment into this job's environment, discarding duplicate
+    # values and ensuring that new ones are prepended.
+    for key, new_values_set in updated_env.items():
+        if key in job_env:
+            prepend_values = new_values_set - set(job_env[key].split(':'))
+            job_env[key] = ':'.join(prepend_values) + ':' + job_env[key]
+        else:
+            job_env[key] = ':'.join(new_values_set)
     return 0
 
 

--- a/catkin_tools/jobs/utils.py
+++ b/catkin_tools/jobs/utils.py
@@ -82,7 +82,8 @@ def loadenv(logger, event_queue, job_env, package, context):
     for key, new_values_set in updated_env.items():
         if key in job_env:
             prepend_values = new_values_set - set(job_env[key].split(':'))
-            job_env[key] = ':'.join(prepend_values) + ':' + job_env[key]
+            if prepend_values:
+                job_env[key] = ':'.join(prepend_values) + ':' + job_env[key]
         else:
             job_env[key] = ':'.join(new_values_set)
     return 0

